pkg/app/piped/apistore/eventstore/store.go

```diff
@@ -0,0 +1,180 @@
+)
+
+// Getter helps get an event. All objects returned here must be treated as read-only.
+type Getter interface {
+	// GetLatest returns the latest event that meets the given conditions.
+	GetLatest(ctx context.Context, name string, labels map[string]string) (*model.Event, bool)
+}
+
+type Store interface {
+	// Run starts syncing the event list with the control-plane.
+	Run(ctx context.Context) error
+	// Getter returns a getter for retrieving an event.
+	Getter() Getter
+}
+
+type apiClient interface {
+	GetLatestEvent(ctx context.Context, req *pipedservice.GetLatestEventRequest, opts ...grpc.CallOption) (*pipedservice.GetLatestEventResponse, error)
+	ListEvents(ctx context.Context, req *pipedservice.ListEventsRequest, opts ...grpc.CallOption) (*pipedservice.ListEventsResponse, error)
+}
+
+type store struct {
+	apiClient    apiClient
+	syncInterval time.Duration
+	gracePeriod  time.Duration
+	logger       *zap.Logger
+
+	// Mark that it has handled all events that was created before this UNIX time.
+	milestone int64
+	mu        sync.RWMutex
+	// The key is supposed to be a string consists of name and labels.
+	// And the value is the address to the latest Event.
+	latestEventMap map[string]*model.Event
+}
+
+const (
+	defaultSyncInterval = time.Minute
+)
+
+// NewStore creates a new event store instance.
+// This syncs with the control plane to keep the list of events for this runner up-to-date.
+func NewStore(apiClient apiClient, gracePeriod time.Duration, logger *zap.Logger) Store {
+	return &store{
+		apiClient:      apiClient,
+		syncInterval:   defaultSyncInterval,
+		gracePeriod:    gracePeriod,
+		latestEventMap: make(map[string]*model.Event),
+		logger:         logger.Named("event-store"),
+	}
+}
+
+// Run starts runner that periodically makes the Events in the cache up-to-date
+// by fetching from the control-plane.
+func (s *store) Run(ctx context.Context) error {
+	s.logger.Info("start running event store")
+
+	syncTicker := time.NewTicker(s.syncInterval)
+	defer syncTicker.Stop()
+
+	// Do first sync without waiting the first ticker.
+	s.milestone = time.Now().Add(-time.Hour).Unix()
+	s.sync(ctx)
+
+	for {
+		select {
+		case <-syncTicker.C:
+			if err := s.sync(ctx); err != nil {
+				s.logger.Error("failed to sync events", zap.Error(err))
+			}
+
+		case <-ctx.Done():
+			s.logger.Info("event store has been stopped")
+			return nil
+		}
+	}
+}
+
+// sync fetches a list of events newly created after its own milestone.
+func (s *store) sync(ctx context.Context) error {
+	resp, err := s.apiClient.ListEvents(ctx, &pipedservice.ListEventsRequest{
+		From:  s.milestone,
+		Order: pipedservice.ListOrder_ASC,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to list events: %w", err)
+	}
+
+	// Make the cache up-to-date by traversing events sorted by oldest first.
+	var latestTime int64
+	s.mu.Lock()
+	for _, event := range resp.Events {
+		id := eventDefinitionID(event.Name, event.Labels)
+		s.latestEventMap[id] = event
```

Comment:
We are doing asynchronous operations so this event can be older than the one stored in the map.
Let's check the CreatedAt before overriding.

Reply:
At the beginning of the loop, it is more likely to happen as you mentioned. But eventually, all Events on the cache will be up-to-date because the list of events is sorted by oldest first. I'd say it's enough but what do you think?

Reply:
Ah, now I could get what you're most worried about. Okay, let me fix them.

Reply:
Yes. The list is updated by `GetLatest` too.